[SE-0046] Implementation for consistent func labels

manavgabhawala opened the pull request:
#### What's in this pull request?
Implements [SE-0046](https://github.com/apple/swift-evolution/blob/master/proposals/0046-first-label.md) to provide consistent function labels (i.e. the first parameter no longer has special behavior for function labels). 

I separated out the changes into separate commits depending on the part being changed so that its easier to review since this is a massive change across the entire repo. However, if it is preferred that this is merged as a single commit, I would be happy to squash into one commit.

I updated most of the swift code using Regex find and replace (and then manual verification where necessary) and since a lot of other repos are going to suffer source code breaking changes to make it easier for them, here are the regex strings I used:
For regular (non-generic) functions I did a search for  `func ([_a-zA-Z0-9]+?)(\s*?)\((\s*?)([a-zA-Z0-9]+?)(\s*?):` and replaced with `func \1\2(\3_ \4\5:` By estimation this worked about 95% of the time and almost never found any false positives.
For generic functions I did a second search using: `func ([_a-zA-Z0-9]+?)(\s*?)(<[\s\S]*?>)(\s*?)\((\s*?)([a-zA-Z0-9]+?)(\s*?):` and replaced with `func \1\2\3\4(\5_ \6\7:` This worked about 90% of the time but be warned it does tend to find false positives (rarely).

#### Resolved bug number:
- [SR-961](https://bugs.swift.org/browse/SR-961)
- rdar://problem/25279084

* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [ ] Review changes
- [ ] Other swift repositories should be ready for a source code breaking change
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please smoke test
        OS X platform               | @swift-ci Please smoke test OS X platform
        Linux platform              | @swift-ci Please smoke test Linux platform

 **Validation Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please test
        OS X platform               | @swift-ci Please test OS X platform
        Linux platform              | @swift-ci Please test Linux platform

Note: Only members of the Apple organization can trigger swift-ci.
</details>
<!-- Thank you for your contribution to Swift! -->

gribozavr commented on stdlib/private/StdlibCollectionUnittest/LoggingWrappers.swift.gyb:
```diff
@@ -139,16 +139,16 @@ public struct LoggingRangeReplaceableCollection<
     C : Collection where C.Iterator.Element == Base.Iterator.Element
   >(
-    bounds: Range<Base.Index>, with newElements: C
+    subRange bounds: Range<Base.Index>, with newElements: C
```
Please don't change the standard library API.  It was already audited and it conforms to the guidelines.

gribozavr commented:
CC @harlanhaskins

gribozavr commented:
@manavgabhawala Thank you!  I am extremely concerned about false positives.  95% is good, but not perfect.  We can't allow any unintended changes in the standard library API.  Did you manually audit all of the changes to the `stdlib` directory?

harlanhaskins commented:
@manavgabhawala Great job! I definitely share @gribozavr's concerns about API -- it's tedious but you'll need to manually verify your regex didn't cause any unintended API changes.

manavgabhawala commented:
@gribozavr @harlanhaskins I did not use the regex for any of the stdlib changes for exactly that reason. I did not change any of the APIs other than the replaceSubrange function (and not just in String, in all the containers), is because almost all of the declarations were of the form replaceSubrange(_ subRange:,with:) so I thought this might be an easier change to do. However if you want, I can go ahead and change all of these back to the original form.

harlanhaskins commented:
I think it's probably best to change it back for this pull request.

harlanhaskins commented on stdlib/public/core/Reflection.swift:
```diff
@@ -497,6 +497,11 @@ struct _EnumMirror : _Mirror {
   subscript(i: Int) -> (String, _Mirror) {
     return _getEnumChild(i, data)
   }
+<<<<<<< dde2aff82c1da3ea293c958931b1a2a7227e86db
```
There are some conflict markers scattered through this file

harlanhaskins commented:
There are some conflict markers scattered throughout `stdlib/core/Reflection.swift`

manavgabhawala commented:
@harlanhaskins will fix, thanks for the catch

manavgabhawala commented:
@harlanhaskins fixed all the changes requested

harlanhaskins commented:
Were those conflicts from abfecfd? Thanks for fixing those so quickly!

manavgabhawala commented:
No I think they were from `[9ce9784180be8ebc7744a7a295c6006f543da344](https://github.com/apple/swift/commit/9ce9784180be8ebc7744a7a295c6006f543da344)`. I had fixed them but I think I forgot to save the file before I rebased. But it should be fixed now. There are always conflicts every time there are new commits since this PR changes a lot of files.

Yeah absolutely, please make sure nothing stands out.

harlanhaskins commented:
We're also going to need to have branches ready to go in SPM, corelibs, lldb, and xctest, because as soon as Top-of-Tree `swiftc` receives this change, all subprojects will fail to build.

I'll coordinate with @mxcl, @parkera, @ddunbar, and @eschaton to get their projects ready.

manavgabhawala commented:
Great, sounds good. Let me know if there's anything I can help with

harlanhaskins commented on stdlib/public/core/ArrayBufferProtocol.swift:
```diff
@@ -69,8 +69,8 @@ public protocol _ArrayBufferProtocol : MutableCollection {
   /// `_ContiguousArrayBuffer`.
-  mutating func replace<C : Collection where C.Iterator.Element == Element>(
-    subRange subRange: Range<Int>,
+  mutating func replaceSubrange<C : Collection where C.Iterator.Element == Element>(
```
Lingering API change here

harlanhaskins commented on stdlib/public/core/ArrayBufferProtocol.swift:
```diff
@@ -134,10 +134,10 @@ extension _ArrayBufferProtocol {
   }
 
-  public mutating func replace<
+  public mutating func replaceSubrange<
```
And here

harlanhaskins commented on docs/SequencesAndCollections.rst:
```diff
@@ -306,10 +306,10 @@ range of elements, denoted by two indices, by elements from a collection with a
 ::
 
   public protocol RangeReplaceableCollectionType : MutableCollectionType {
-    mutating func replaceSubrange<
```
Another here

harlanhaskins commented on docs/StdlibAPIGuidelines.rst:
```diff
@@ -88,7 +88,7 @@ Subsequent Parameters
   .. parsed-literal::
 
-    x.replaceSubrange(r, **with:** someElements)
+	x.replace(subRange: r, **with:** someElements)
```
And here

harlanhaskins commented on docs/archive/LangRefNew.rst:
```diff
@@ -1348,7 +1348,7 @@ otherwise empty cases in switch statements.
 
 ::
 
-  func classifyPoint(point: (Int, Int)) {
```
API change

gribozavr replied:
Is it?

harlanhaskins replied:
It's not actually a change in API, but I don't think changing the doc for API naming is in scope for this change.

Though I'm not against this change if you wanna just take it.

harlanhaskins commented on docs/archive/Objective-CInteroperability.rst:
```diff
@@ -496,21 +496,21 @@ Overloading
 
     // 1. foo:baz:
-    func foo(Int bar, Int baz);
+    func foo(_ bar: Int, baz: Int)
```
I applaud you for finding and fixing these.

@gribozavr What do you think about these changes? It's an archived document with old syntax.

gribozavr replied:
Agreed -- I'd rather avoid making further edits to this document.

manavgabhawala replied:
@gribozavr so should I get rid of these changes then, or leave them in?

gribozavr replied:
Let's revert the changes from this file.

harlanhaskins commented on docs/archive/Objective-CInteroperability.rst:
```diff
@@ -496,21 +496,21 @@ Overloading
     // 2. foo:qux:
-    func foo(Int bar, Int qux);
-  
+    func foo(_ bar: Int bar, qux: Int)
```
`Int bar` repeated here

harlanhaskins commented on docs/proposals/OptimizerEffects.rst:
```diff
@@ -185,17 +185,17 @@ state.
       var storage: ArrayStorage
 
       @preserve_unique
-      mutating func replaceSubrange<
```
API change

harlanhaskins replied:
(Even though these are just docs)

harlanhaskins commented on docs/proposals/ValueSemantics.rst:
```diff
@@ -179,7 +179,7 @@ Here's a version of cycle_length that works when state is a mutable
 
  func cycle_length<State>(
-   s : State, mutate : ( [inout] State ) -> () 
+   + s : State, mutate : ( [inout] State ) -> ()
```
`+` instead of `_`

harlanhaskins commented on stdlib/private/StdlibCollectionUnittest/CheckRangeReplaceableCollectionType.swift:
```diff
@@ -75,7 +75,7 @@ internal struct ReplaceRangeTest {
     self.newElements = newElements.map(OpaqueValue.init)
     self.rangeSelection = rangeSelection
     self.expected = expected
-    self.loc = SourceLoc(file, line, comment: "replaceSubrange() test data")
```
replaceSubrange -> replace

harlanhaskins commented on stdlib/private/StdlibCollectionUnittest/MinimalCollections.swift.gyb:
```diff
@@ -1134,11 +1134,11 @@ public struct ${Self}<T> : RangeReplaceableCollection {
     let oldCount = count
     elements.replaceSubrange(
-      subRange.startIndex.position..<subRange.endIndex.position,
+	  subRange.startIndex.position..<subRange.endIndex.position,
```
nit-pick: Indentation

harlanhaskins commented:
I can't see the file right now, but whilst running tests I found a conflict marker in `test/expr/unary/selector/selector.swift`

harlanhaskins commented on stdlib/public/core/ArrayBuffer.swift:
```diff
@@ -392,8 +391,8 @@ extension _ArrayBuffer {
         var refCopy = self
-        refCopy.replace(
-          subRange: i...i, with: 1, elementsOf: CollectionOfOne(newValue))
+        refCopy.replaceSubrange(
```
API change here

harlanhaskins commented:
Notes for remaining files (GitHub's diff tool can't handle this right now)

### Standard Library

- `stdlib/core/SliceBuffer.swift`
  - `replace` -> `replaceSubrange` (line 80)
  - `replace` -> `replaceSubrange` (line 103)
  - `replace` -> `replaceSubrange` (line 152)
- `stdlib/core/String.swift`
  - `replaceSubrange` argument name changed (line 686)
- `stdlib/core/StringCharacterView.swift`
  - `replaceSubrange` argument name changed (line 260)

manavgabhawala commented:
@harlanhaskins I fixed all of these so far. I'm rerunning all the tests before I push the new commit though (just to be sure I didn't accidentally break anything). 
Of course :) 
Don't worry, I'm glad you are nit-picky, its absolutely necessary to keep this massive code base in order!

manavgabhawala commented:
@harlanhaskins I just fixed all the changes requested so far, I don't think I did but, hopefully I didn't forget any.

manavgabhawala commented:
@harlanhaskins fixed `classifyPoint`

harlanhaskins commented:
Small change; almost positively inconsequential, the changes in `test/DebugInfo/AlignedAt.swift` change around some of the arguments, which changes the mangled name. You updated the test for it, but it might be best just to change the declarations.

@gribozavr Thoughts?

manavgabhawala commented:
In some places I changed the declaration and in others I changed the mangled names to test for both scenarios. But if you think it's better to just change the declaration I can fix that

harlanhaskins commented:
That'd be best. Also thanks for changing the replace/replaceSubrange things. Those look good now 👍

harlanhaskins commented:
Hmm...what happened with `test/NameBinding/library.swift`? Were these tests broken before the change?

manavgabhawala commented:
For `test/NameBinding/library.swift`, it used to give us that 'wrong' and weird diagnostic, but with these changes they no longer happen, so I updated the test case to reflect that. Should I have done something else?

harlanhaskins commented:
Nope, I think that's the correct thing to do. @gribozavr? I'd like to know why the same warning didn't show up since we just changed the declaration such that it would parse the same as before.

manavgabhawala commented:
Also, what are you sure you meant `test/DebugInfo/AlignedAt.swift` I couldn't find any test case like that?

harlanhaskins commented:
Oops, sorry, I meant `test/DebugInfo/inlinedAt.swift`

gribozavr commented:
@manavgabhawala I think you need to adjust `test/NameBinding/Inputs/imported_module.swift` to revert to old diagnostics in `test/NameBinding/library.swift`.  This is important since with this change the `over1` functions no longer have the same labels.

manavgabhawala commented:
@gribozavr @harlanhaskins fixed both `test/DebugInfo/inlinedAt.swift` and `test/NameBinding/library.swift `

manavgabhawala commented:
@harlanhaskins I think you meant `test/SILGen/coverage_closures.swift` not `test/SILGen/coverage_exceptions.swift`, I fixed the closures one.
Also `test/expr/expressions.swift` there's no change on line 758, you probably meant another file or a different line? 
Fixed all the others, except for `test/SourceKit/DocSupport/doc_swift_module.swift.response` and `validation-test/SIL/crashers/017-swift-decl-walk.sil`, because I'm not sure what's wrong or what to fix.

harlanhaskins commented:
For `test/expr/expressions.swift`: sorry, it was 798.
I don't know what to do for `doc_swift_module.swift.response`, but my instinct is to remove the diagnostic test from this commit and commit it later once we know what to do.

manavgabhawala commented:
I fixed, `test/expr/expressions.swift`, and pushed a new commit but github is being weird and not showing the new commit for some reason and when I `git push` it says everything up to date. For `doc_swift_module.swift.response `, if I revert this test case to the original, it will fail.

gribozavr commented:
If `validation-test/SIL/crashers/017-swift-decl-walk.sil` does not crash anymore, then it is an improvement!  Please adjust the RUN lines so that we expect a clean exit and move the test to `validation-test/SIL/crashers_fixed/`.

harlanhaskins commented:
@nkcsgexi Any ideas how @manavgabhawala should add a test for a new diagnostic message in `test/SourceKit/DocSupport/doc_swift_module.swift.response`?

nkcsgexi commented:
The test failure shows us two files are different  doc_swift_module.swift.response and doc_swift_module.swift.response.tmp. What I typically do is (1) manually verify the diff is reflects desirable outcomes, and (2) update  doc_swift_module.swift.response to *.tmp, so that they are identical.

harlanhaskins commented:
I think the problem is that his diagnostic includes some absolute file paths, which fails on anyone else's machine...

nkcsgexi commented:
oh. Do we need to add an absolute file path into diagnosis? I think file names will suffice: e.g.
 StringRef FileName = llvm::sys::path::filename(FullPath); // Give you the file name from an absolute path.

harlanhaskins commented:
Sorry, I wasn't specific. This PR isn't generating filenames. What's happening is @manavgabhawala's `doc_swift_module.swift.response` file is generated by the existing code with extra bits appended to the end, specifically:

```
[
  {
    key.line: 81,
    key.column: 17,
    key.filepath: "/Users/manavgabhawala/Documents/Xcode/Swift/swift/test/SourceKit/DocSupport/Inputs/main.swift",
    key.severity: source.diagnostic.severity.error,
    key.description: "method 'protMeth(a:)' has different argument names from those required by protocol 'Prot' ('protMeth')",
    key.fixits: [
      {
        key.offset: 1092,
        key.length: 0,
        key.sourcetext: "_ "
      }
    ],
    key.diagnostics: [
      {
        key.line: 35,
        key.column: 8,
        key.filepath: "/Users/manavgabhawala/Documents/Xcode/Swift/swift/test/SourceKit/DocSupport/Inputs/main.swift",
        key.severity: source.diagnostic.severity.note,
        key.description: "requirement 'protMeth' declared here"
      }
    ]
  }
]
```

and we don't know why.

harlanhaskins commented:
Also, @manavgabhawala `test/SILOptimizer/array_mutable_assertonly.swift` has a `func use(a:` that needs to be changed.

harlanhaskins commented:
Wait, no, I see the issue here. @manavgabhawala You need to change all of the `protMeth(a:)`'s inside `test/SourceKit/DocSupport/Inputs/main.swift`. `class BC2` and `class SubC2` both have incorrect labels.

Sorry to bother you, @nkcsgexi.

nkcsgexi commented:
I see. Thanks for the explanation! When generating doc_swift_module.swift.response, one can assume that the module compiles without any warning message. So the missing part is to update test/SourceKit/DocSupport/Inputs/main.swift and test/SourceKit/DocSupport/Inputs/cake.swift so that no diagnosis is generated.

harlanhaskins commented:
Also @manavgabhawala (I keep finding more things, running the full test suite on OS X)
- `test/1_stdlib/UIKit.swift` needs updating
  - `printDevice(o:)` -> `printDevice(_ o:)`
  - `printInterface(o:)` -> `printInterface(_ o:)`

manavgabhawala commented:
I ran and tested all with `utils/build-script -R -T` Should I test with a different command? Also, the `test/SILOptimizer/array_mutable_assertonly.swift ` was unsupported for me for some reason that's why I didn't catch it. Fixed it though. 
Also updated `test/SourceKit/DocSupport/Inputs/main.swift`, not sure if I need to change `doc_swift_module.swift.response` at all or not? But the test case does pass when I run it.
Also fixed `UIKit.swift`, which was unsupported when I ran it.

manavgabhawala commented:
As for `validation-test/SIL/crashers/017-swift-decl-walk.sil `, I moved it to the right directory and removed the `--crash` from the file.

gribozavr commented:
@swift-ci Please test

harlanhaskins commented:
@gribozavr Oh no; we can't test this without the changes in SPM...

harlanhaskins commented:
@manavgabhawala Also it looks like the extra diagnostics didn't get deleted from `doc_swift_module.swift.response`.

gribozavr commented:
@harlanhaskins I know.  I hoped to get at least some CI feedback.

gribozavr commented:
And we have some:

```
/Users/buildnode/jenkins/workspace/swift-PR-osx/swift/stdlib/public/core/ObjCMirrors.swift:46:41: error: missing argument label 'object:' in call
    return _getClassPlaygroundQuickLook(object)
```

harlanhaskins commented:
Interesting! I didn't get that one when I built with `--test --validation-test`...

harlanhaskins commented:
Oh, these are new changes from master?

manavgabhawala commented:
@harlanhaskins can I rebase now, if you are done reviewing? so that I can fixup the new extra commits that I pushed and also get all the new changes to master?

harlanhaskins commented:
Yes, absolutely! Thanks again!

We're going to coordinate this with corelibs-foundation, XCTest, and SPM, to get all of these merged together at the same time.

There will undoubtedly be at least one more rebase in your future.

shahmishal commented:
@swift-ci Please test

shahmishal commented:
@swift-ci Please smoke test

harlanhaskins commented:
Smoke test passed. Are we set to merge?

harlanhaskins commented:
Some conflicts just arose 😕

@manavgabhawala Mind fixing these and squashing into one commit?

harlanhaskins commented:
You just need to delete the definitions for `insetBy` and the like, inside `stdlib/core/SDK/CoreGraphics/CoreGraphics.swift`.

manavgabhawala commented:
@harlanhaskins do you want just one commit total, or squash into the current 5, correctly?

harlanhaskins commented:
@manavgabhawala One commit total, if possible.

manavgabhawala commented:
@harlanhaskins done. Should we re- smoke test?

harlanhaskins commented:
@manavgabhawala I don't think that will be necessary. @gribozavr?

harlanhaskins commented:
@swift-ci Please smoke test

harlanhaskins commented:
Thank you, @manavgabhawala!
